src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,617 @@
+            for endpoint in read_request:
+                nested_output = read_request[endpoint]
+                for cluster in nested_output:
+                    attributes = nested_output[cluster]
+                    asserts.assert_true(isinstance(attributes[chosen_cluster], cluster_type),
+                                        f"{chosen_cluster} is not a {cluster_type}")
+        return
+
+    @async_test_body
+    async def test_TC_IDM_2_2(self):
+
+        self.user_params["use_pase_only"] = False
+        super().setup_class()
+        await self.setup_class_helper()
+        # Test Setup
+
+        all_clusters = [cluster for cluster in Clusters.ClusterObjects.ALL_ATTRIBUTES]
+
+        server_list_attr = Clusters.Objects.Descriptor.Attributes.ServerList
+        server_list_attr_path = [(0, server_list_attr)]
+        descriptor_obj = Clusters.Objects.Descriptor
+        descriptor_obj_path = [(0, descriptor_obj)]
+        attribute_list = Clusters.Objects.Descriptor.Attributes.AttributeList
+        attribute_list_path = [0, attribute_list]
+        self.all_supported_clusters = [cluster for cluster in Clusters.__dict__.values(
+        ) if inspect.isclass(cluster) and issubclass(cluster, ClusterObjects.Cluster)]
+
+        # # For str test
+        # power_source = Clusters.Objects.PowerSource
+
+        # # For uint test
+        # microwave_oven_mode = Clusters.Objects.MicrowaveOvenMode
+
+        # # For int test
+        # device_energy_mgmt = Clusters.Objects.DeviceEnergyManagement
+
+        # # For float test
+        # unit_testing = Clusters.Objects.UnitTesting
+
+        # # For list test
+        # energy_evse_mode = Clusters.Objects.EnergyEvseMode
+
+        # # For bitmap test
+        # occupancy_sensing = Clusters.Objects.OccupancySensing
+
+        self.print_step(0, "Commissioning - already done")
+
+        wildcard_descriptor = await self.default_controller.ReadAttribute(self.dut_node_id, [(Clusters.Descriptor)])
+        endpoints = list(wildcard_descriptor.keys())
+
+        endpoints.sort()
+        # non_existent_endpoint = next(i for i, e in enumerate(endpoints + [None]) if i != e)
+
+        # Step 1
+
+        # TH sends the Read Request Message to the DUT to read one attribute on a given cluster and endpoint.
+        # AttributePath = [[Endpoint = Specific Endpoint, Cluster = Specific ClusterID, Attribute = Specific Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value to the DUT
+
+        self.print_step(1, "Send Request Message to read one attribute on a given cluster and endpoint")
+
+        # endpoint = [(0, Clusters.Objects.Descriptor.Attributes.ServerList)]
+        read_request_1 = await self.default_controller.ReadAttribute(self.dut_node_id, server_list_attr_path)
+        returned_endpoints = read_request_1[0].keys()
+        # returned_endpoints = dict_keys([<class 'chip.clusters.Objects.Descriptor'>])
```

Comment:
Removal will show up in next commit (the first comment was actually an aid to indicate the value of endpoint so that the user didn't have to scroll up)

Reply:
Removed in a1c592a